File: lib-util/src/main/java/com/ustadmobile/core/impl/UmLifecycleOwner.java
package com.ustadmobile.core.impl;

/**
 * Represents an object (e.g. presenter) that has a lifecycle.
 */
public interface UmLifecycleOwner {

    int STATUS_CREATED = 1;

    int STATUS_STARTED = 2;

    int STATUS_RESUMED = 3;

    int STATUS_PAUSED = 4;

    int STATUS_STOPPED = 5;

    int STATUS_DESTROYED = 6;

    /**
     * Get the system context object
     *
     * @return the system context object
     */
    Object getContext();

    /**
     * Add an event listener for lifecycle changes
     *
     * @param listener event listener
     */
    void addLifecycleListener(UmLifecycleListener listener);

    /**
     * Remove an event listener for lifecycle changes
     *
     * @param listener event listener
     */
    void removeLifecycleListener(UmLifecycleListener listener);

    int getCurrentStatus();

}

File: lib-util/src/main/java/com/ustadmobile/core/impl/UmLifecycleOwner.kt
package com.ustadmobile.core.impl

/**
 * Represents an object (e.g. presenter) that has a lifecycle.
 */
interface UmLifecycleOwner {

    /**
     * Get the system context object
     *
     * @return the system context object
     */
    val context: Any

    /**
     * Add an event listener for lifecycle changes
     *
     * @param listener event listener
     */
    fun addLifecycleListener(listener: UmLifecycleListener)

    /**
     * Remove an event listener for lifecycle changes
     *
     * @param listener event listener
     */
    fun removeLifecycleListener(listener: UmLifecycleListener)

    companion object {

        const val STATUS_CREATED = 1

        const val STATUS_STARTED = 2

        const val STATUS_RESUMED = 3

        const val STATUS_PAUSED = 4

        const val STATUS_STOPPED = 5

        const val STATUS_DESTROYED = 6
    }

}
